rasa/core/policies/rule_policy.py

```diff
@@ -212,15 +242,137 @@ def _create_loop_unhappy_lookup_from_states(
+                return
+
+        tracker_type = "rule" if tracker.is_rule_tracker else "story"
+        return (
+            f"- the prediction of the action '{gold_action_name}' in {tracker_type} "
+            f"'{tracker.sender_id}' is contradicting with another rule or story."
+        )
+
+    def _find_contradicting_rules(
+        self,
+        trackers: List[TrackerWithCachedStates],
+        domain: Domain,
+        interpreter: NaturalLanguageInterpreter,
+    ) -> None:
+        logger.debug("Started checking rules and stories for contradictions.")
+        # during training we run `predict_action_probabilities` to check for
+        # contradicting rules.
+        # We silent prediction debug to avoid too many logs during these checks.
+        logger_level = logger.level
+        logger.setLevel(logging.WARNING)
+        error_messages = []
+        pbar = tqdm(
+            trackers,
+            desc="Processed trackers",
+            disable=rasa.shared.utils.io.is_logging_disabled(),
+        )
+        for tracker in pbar:
+            running_tracker = tracker.init_copy()
+            running_tracker.sender_id = tracker.sender_id
+            # the first action is always unpredictable
+            next_action_is_unpredictable = True
+            for event in tracker.applied_events():
+                if not isinstance(event, ActionExecuted):
+                    running_tracker.update(event)
+                    continue
+
+                if event.action_name == RULE_SNIPPET_ACTION_NAME:
+                    # notify that we shouldn't check that the action after
+                    # RULE_SNIPPET_ACTION_NAME is unpredictable
+                    next_action_is_unpredictable = True
+                    # do not add RULE_SNIPPET_ACTION_NAME event
+                    continue
+
+                # do not run prediction on unpredictable actions
+                if next_action_is_unpredictable or event.unpredictable:
+                    next_action_is_unpredictable = False  # reset unpredictability
+                    running_tracker.update(event)
+                    continue
+
+                gold_action_name = event.action_name or event.action_text
+                error_message = self._check_prediction(
+                    running_tracker, domain, interpreter, gold_action_name
+                )
+                if error_message:
+                    error_messages.append(error_message)
+
+                running_tracker.update(event)
+
+        logger.setLevel(logger_level)  # reset logger level
+        if error_messages:
+            error_messages = "\n".join(error_messages)
+            raise InvalidRule(
+                f"\nContradicting rules 🚨\n\n{error_messages}\n"
+                f"Please update your stories and rules so that they don't contradict "
+                f"each other."
+            )
+
+        logger.debug("Found no contradictions between rules and stories")
```

Comment:
should it say "Found no contradicting rules"? since rules can also contradict with other rules